#VRML_SIM R2025a utf8

EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/objects/backgrounds/protos/TexturedBackground.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/objects/backgrounds/protos/TexturedBackgroundLight.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/objects/factory/containers/protos/WoodenBox.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/objects/floors/protos/Floor.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/appearances/protos/Parquetry.proto"

WorldInfo {
  basicTimeStep 16
}
Viewpoint {
  orientation -0.1698175282290138 -0.8835365462698728 0.43649189970913876 5.453629326564267
  position -0.8751716023350382 0.2886546888070706 1.2101696588269735
}
Background {
  skyColor [
    0.5 0.7 1
  ]
}
TexturedBackground {
}
TexturedBackgroundLight {
}
Floor {
  size 3 3
  appearance Parquetry {
    type "chequered"
    colorOverride 1 0.666667 1
  }
}
WoodenBox {
  translation 0.626188 -0.992046 0.15
  size 0.3 0.3 0.3
}
WoodenBox {
  translation 0.849962 0.952797 0.15
  name "wooden box with mass"
  size 0.3 0.3 0.3
  mass 0.2
}
WoodenBox {
  translation -0.607032 0.652461 0.15
  name "wooden box(1)"
  size 0.3 0.3 0.3
}
Solid {
  translation 0.2 0.44 0.2
  children [
    Shape {
      appearance DEF WALL_APPEARANCE PBRAppearance {
        roughness 1
        metalness 0
      }
      geometry DEF BALL_GEOMETRY Sphere {
        radius 0.1
      }
    }
  ]
  name "Ballothy"
  boundingObject USE BALL_GEOMETRY
  physics Physics {
  }
}
Solid {
  translation 1.45 0 0.05
  children [
    Shape {
      appearance DEF WALL_APPEARANCE PBRAppearance {
        baseColor 0 0.333333 1
      }
      geometry DEF WALL_GEOMETRY Box {
        size 0.1 3 0.1
      }
    }
  ]
  name "Wall 1"
  boundingObject USE WALL_GEOMETRY
}
Solid {
  translation -1.45 0 0.05
  children [
    Shape {
      appearance USE WALL_APPEARANCE
      geometry USE WALL_GEOMETRY
    }
  ]
  name "Wall 2"
  boundingObject USE WALL_GEOMETRY
}
Solid {
  translation 0 1.45 0.05
  rotation 0 0 1 1.5708
  children [
    Shape {
      appearance USE WALL_APPEARANCE
      geometry USE WALL_GEOMETRY
    }
  ]
  name "Wall 3"
  boundingObject USE WALL_GEOMETRY
}
Solid {
  translation 0 -1.45 0.05
  rotation 0 0 1 1.5708
  children [
    Shape {
      appearance USE WALL_APPEARANCE
      geometry USE WALL_GEOMETRY
    }
  ]
  name "Wall 4"
  boundingObject USE WALL_GEOMETRY
}
Robot {
  translation 0 0 0.06
  children [
    DistanceSensor {
      translation 0.1 0.025 0
      rotation 0 0 1 0.3
      children [
        DEF sens_shape Shape {
          appearance USE WALL_APPEARANCE
          geometry DEF SENS_BOX Box {
            size 0.01 0.01 0.01
          }
          castShadows FALSE
        }
      ]
      name "ds_left"
      boundingObject USE SENS_BOX
    }
    DistanceSensor {
      translation 0.1 -0.025 0
      rotation 0 0 1 -0.3
      children [
        USE sens_shape
      ]
      name "ds_right"
      boundingObject USE SENS_BOX
    }
    Shape {
      appearance PBRAppearance {
        baseColor 1 0 0
      }
      geometry DEF _WHEELBODY Box {
        size 0.2 0.1 0.05
      }
    }
    HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 1 0
        anchor 0.05 0.06 0
      }
      device [
        RotationalMotor {
          name "wheel1"
        }
      ]
      endPoint DEF WHEEL1 Solid {
        translation 0.05 0.06 0
        rotation 1 0 0 1.5707996938995747
        children [
          Shape {
            geometry DEF WHEEL Cylinder {
              height 0.02
              radius 0.04
            }
          }
        ]
        boundingObject USE WHEEL
        physics DEF WHEEL_PH Physics {
        }
      }
    }
    HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 1 0
        anchor -0.05 0.06 0
      }
      device [
        RotationalMotor {
          name "wheel3"
        }
      ]
      endPoint DEF WHEEL2 Solid {
        translation -0.05 0.06 0
        rotation 1 0 0 1.5707996938995747
        children [
          Shape {
            geometry USE WHEEL
          }
        ]
        name "solid(1)"
        boundingObject USE WHEEL
        physics USE WHEEL_PH
      }
    }
    HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 1 0
        anchor 0.05 -0.06 0
      }
      device [
        RotationalMotor {
          name "wheel2"
        }
      ]
      endPoint DEF WHEEL3 Solid {
        translation 0.05 -0.06 0
        rotation 1 0 0 1.5707996938995747
        children [
          Shape {
            geometry USE WHEEL
          }
        ]
        name "solid(2)"
        boundingObject USE WHEEL
        physics Physics {
        }
      }
    }
    HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 1 0
        anchor -0.05 -0.06 0
      }
      device [
        RotationalMotor {
          name "wheel4"
        }
      ]
      endPoint DEF WHEEL4 Solid {
        translation -0.05 -0.06 0
        rotation 1 0 0 1.5707996938995747
        children [
          Shape {
            geometry USE WHEEL
          }
        ]
        name "solid(3)"
        boundingObject USE WHEEL
        physics Physics {
        }
      }
    }
  ]
  boundingObject USE _WHEELBODY
  physics Physics {
  }
  controller "four_wheel_collision_avoidance"
}
